File: python-worker/worker.py
#!/usr/bin/env python3
import os
import time
import requests
import sentry_sdk
from sentry_sdk.integrations.logging import LoggingIntegration
from dotenv import load_dotenv

load_dotenv()
print(f"[Python Worker] DSN loaded: {'YES' if os.getenv('SENTRY_DSN') else 'NO'}")

def before_send(event, hint):
    event_type = event.get('type')
    if event_type == 'transaction' or 'transaction' in event or 'spans' in event:
        if 'sampled' not in event or event.get('sampled') is not True:
            event['sampled'] = True
    return event

sentry_sdk.init(
    dsn=os.getenv("SENTRY_DSN"),
    traces_sample_rate=1.0,
    environment="development",
    debug=True,
    integrations=[
        LoggingIntegration(level=None, event_level=None),
    ],
    before_send=before_send,
)

def process_message(message):
    sentry_trace = message.get('sentryTrace')
    baggage = message.get('baggage')
    
    if sentry_trace:
        headers = {'sentry-trace': sentry_trace}
        if baggage:
            headers['baggage'] = baggage
        
        try:
            trace_envelope = sentry_sdk.continue_trace(headers)
            transaction = sentry_sdk.start_transaction(trace_envelope)
            
            with transaction:
                with sentry_sdk.start_span(
                    op='queue.process',
                    description='python-worker-processing'
                ) as span:
                    time.sleep(0.5)
                    span.set_tag('task.type', message.get('taskType', 'unknown'))
                    span.set_status('ok')
            
            return {'success': True, 'processedBy': 'python-worker'}
        except Exception as e:
            print(f"[Python Worker] Error: {e}")
            sentry_sdk.capture_exception(e)
            return {'success': False, 'error': str(e)}
    else:
        print("[Python Worker] No trace context")
        return {'success': False, 'error': 'No trace context'}

QUEUE_API_URL = os.getenv('QUEUE_API_URL', 'http://localhost:3002')

print(f"[Python Worker] Starting worker, polling {QUEUE_API_URL}")

while True:
    try:
        response = requests.post(
            f'{QUEUE_API_URL}/queue/receive',
            json={'queueName': 'python-worker-queue', 'maxMessages': 1},
            timeout=5
        )
        
        if response.status_code == 200:
            data = response.json()
            messages = data.get('messages', [])
            
            for message in messages:
                try:
                    result = process_message(message)
                    print(f"[Python Worker] Processed: {result}")
                except Exception as e:
                    print(f"[Python Worker] Error processing message: {e}")
                    sentry_sdk.capture_exception(e)
        
        time.sleep(1)
    except Exception as e:
        print(f"[Python Worker] Polling error: {e}")
        time.sleep(5)

File: node-worker/worker.js
import * as Sentry from '@sentry/node';
import { nodeProfilingIntegration } from '@sentry/profiling-node';
import dotenv from 'dotenv';

dotenv.config();

console.log('[Node Worker] DSN loaded:', process.env.SENTRY_DSN ? 'YES' : 'NO');

Sentry.init({
  dsn: process.env.SENTRY_DSN,
  integrations: [
    nodeProfilingIntegration(),
  ],
  tracesSampleRate: 1.0,
  profilesSampleRate: 1.0,
  environment: 'development',
});

async function processMessage(message) {
  console.log('[Node Worker] Processing message:', {
    messageId: message.MessageId,
    sentryTrace: message.sentryTrace,
    hasTrace: !!message.sentryTrace,
  });

  if (!message.sentryTrace) {
    console.warn('[Node Worker] No sentryTrace in message');
    return { success: false, error: 'No trace context' };
  }

  return Sentry.continueTrace(
    { 
      sentryTrace: message.sentryTrace, 
      baggage: message.baggage 
    },
    () => {
      return Sentry.startSpan(
        {
          name: 'node-worker-transaction',
          op: 'function',
        },
        async (parentSpan) => {
          return await Sentry.startSpan(
            {
              name: 'node-worker.process',
              op: 'queue.process',
              attributes: {
                'messaging.message.id': message.MessageId,
                'messaging.destination.name': 'task-queue',
                'messaging.message.body.size': JSON.stringify(message).length,
                'task.type': message.taskType,
              },
            },
            async (span) => {
              await new Promise(resolve => setTimeout(resolve, 100));
              
              const traceHeader = Sentry.spanToTraceHeader(span);
              const baggageHeader = Sentry.spanToBaggageHeader(span);
              
              await fetch('http://localhost:3002/queue/send', {
                method: 'POST',
                headers: {
                  'Content-Type': 'application/json',
                  'sentry-trace': traceHeader,
                  'baggage': baggageHeader || '',
                },
                body: JSON.stringify({
                  queueName: 'python-worker-queue',
                  message: {
                    ...message,
                    sentryTrace: traceHeader,
                    baggage: baggageHeader,
                  },
                }),
              });
              
              parentSpan.setStatus({ code: 1, message: 'ok' });
              await Sentry.flush(2000);
              
              return { success: true, processedBy: 'node-worker' };
            }
          );
        }
      );
    }
  );
}

async function pollQueue() {
  while (true) {
    try {
      const response = await fetch('http://localhost:3002/queue/receive', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          queueName: 'task-queue',
          maxMessages: 1,
        }),
      });
      
      const data = await response.json();
      
      if (data.messages && data.messages.length > 0) {
        for (const message of data.messages) {
          try {
            await processMessage(message);
          } catch (error) {
            console.error('[Node Worker] Error processing message:', error);
            Sentry.captureException(error);
          }
        }
      }
      
      await new Promise(resolve => setTimeout(resolve, 1000));
    } catch (error) {
      console.error('[Node Worker] Error polling queue:', error);
      await new Promise(resolve => setTimeout(resolve, 5000));
    }
  }
}

console.log('[Node Worker] Starting worker...');
pollQueue();
